### ISSUE TITLE
Schema-aware query analysis

### ISSUE THREAD

**Issue by rnewman** [mozilla/datomish#14]
When the domain and range of an attribute are known (and in our case the domain is always an entity), we can do tiers of analysis on queries.

The first tier is trivial value-space comparison. A pattern like:

```
[?foo :bar/baz 5.0]
```

where `:bar/baz` is defined as `valueType` `ref` is obviously going to fail to return any bindings (with the obvious exception of any collision in our design between integer ref IDs and integer values). That's very likely a programming error, and we should warn as such, as well as optimizing the query.

The second tier is to do this across multiple patterns:

```
[?foo :bar/age ?age]
[?age :bar/name ?name]
```

expects `?age` to be both a ref and a number, and thus cannot succeed.

**Comment by rnewman** [comment]
Beyond this one gets into real description logic stuff, like disjoint classes for attribute domains, which is probably more than we need.

**Comment by rnewman** [comment]
This is working in my branch.

**Comment by rnewman** [comment]
Fixed in b2a1af30ed6f13ca72012a317b81a5a8127ecf3f.
